test/python/dagcircuit/test_dagcircuit.py

```diff
@@ -248,6 +248,140 @@ def test_addding_individual_bit(self):
+        self.creg1 = creg1
+        self.creg1_alias = creg2
+        self.individual_clbit = clbit
+
+        self.original_cregs = [self.creg0, self.creg1, self.creg1_alias]
+
+        # skip adding bits of creg1_alias since it's just an alias for creg1
+        self.original_clbits = [
+            b for reg in self.original_cregs if reg != self.creg1_alias for b in reg
+        ] + [self.individual_clbit]
+
+    def assert_cregs_is(self, cregs, excluding=None):
+        """Assert test DAG cregs match the expected values.
+
+        Args:
+            cregs (Iterable(ClassicalRegister)): the classical registers to expect
+            excluding (Set(ClassicalRegister)): classical registers to remove from
+            ``cregs`` before the comparison."""
+        if excluding is None:
+            excluding = {}
+        self.assertDictEqual(
+            self.dag.cregs, {creg.name: creg for creg in cregs if creg not in excluding}
+        )
+
+    def assert_clbits_is(self, clbits, excluding=None):
+        """Assert test DAG clbits match the expected values.
+
+        Args:
+            clbits (Iterable(Clbit)): the classical bits to expect
+            excluding (Set(ClassicalRegister)): classical bits to remove from
+            ``clbits`` before the comparison."""
+        if excluding is None:
+            excluding = {}
+        self.assertListEqual(self.dag.clbits, list(b for b in clbits if b not in excluding))
+
+    def test_remove_idle_creg(self):
+        """Removing an idle classical register removes just the register."""
+        self.dag.remove_cregs(self.creg0)
+
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg0})
+        self.assert_clbits_is(self.original_clbits)
+
+    def test_remove_busy_creg(self):
+        """Classical registers with both busy and idle underlying bits
+        can be removed, keeping underlying bits."""
+        self.dag.apply_operation_back(Measure(), [self.qreg[0]], [self.creg0[0]])
+        self.dag.remove_cregs(self.creg0)
+
+        # creg removal always succeeds
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg0})
+
+        # creg0 is gone, leaving both creg0[0] and creg0[1] unreferenced.
+        # however, only creg0[1] is idle, and thus removed.
+        self.assert_clbits_is(self.original_clbits)
+
+    def test_remove_cregs_shared_bits(self):
+        """Removing a classical register does not remove other classical
+        registers that are simply aliases of it."""
+        self.dag.remove_cregs(self.creg1)
+
+        # Only creg1 should be deleted, not its alias
+        self.assert_cregs_is(self.original_cregs, excluding={self.creg1})
+        self.assert_clbits_is(self.original_clbits)
+
+    def test_remove_unknown_creg(self):
+        """Classical register removal of unknown registers raises."""
+        unknown_creg = ClassicalRegister(1)
+        with self.assertRaises(DAGCircuitError):
+            self.dag.remove_cregs(unknown_creg)
```

Comment:
Ideally this would be `self.assertRaisesRegex(DAGCircuitError, ".*small portion of message.*")`, so we're testing that the correct error handling is being triggered.  (etc for other tests)

Reply:
Done in f717f58.